File: backend/app/routers/work_records.py
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from ..auth import get_current_admin_user, get_current_user
from ..database import get_db
from ..models import MonthlyClosing, WorkRecord
from ..schemas import WorkRecordCreate, WorkRecordResponse, WorkRecordUpdate
from ..services.calculation import calculate_actual_minutes

router = APIRouter()


def _assert_month_open(db: Session, user_id: int, year_month: str, is_admin: bool):
    """Raise 403 if the month is closed and the caller is not admin."""
    if is_admin:
        return
    closing = (
        db.query(MonthlyClosing)
        .filter(
            MonthlyClosing.user_id == user_id,
            MonthlyClosing.year_month == year_month,
        )
        .first()
    )
    if closing and closing.status == "closed":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="この月はすでに締め処理済みです",
        )


@router.get("", response_model=List[WorkRecordResponse])
def list_work_records(
    year_month: Optional[str] = Query(None, description="YYYY-MM"),
    user_id: Optional[int] = Query(None),
    db: Session = Depends(get_db),
    current_user=Depends(get_current_user),
):
    """
    List work records.
    - Admins can filter by any user_id (or see all).
    - Employees only see their own records.
    """
    query = db.query(WorkRecord)

    if not current_user.is_admin:
        query = query.filter(WorkRecord.user_id == current_user.id)
    elif user_id is not None:
        query = query.filter(WorkRecord.user_id == user_id)

    if year_month:
        # Filter by YYYY-MM prefix using work_date
        year, month = year_month.split("-")
        from sqlalchemy import extract
        query = query.filter(
            extract("year", WorkRecord.work_date) == int(year),
            extract("month", WorkRecord.work_date) == int(month),
        )

    return query.order_by(WorkRecord.work_date).all()


@router.post("", response_model=WorkRecordResponse, status_code=status.HTTP_201_CREATED)
def create_work_record(
    payload: WorkRecordCreate,
    db: Session = Depends(get_db),
    current_user=Depends(get_current_user),
):
    """Create a work record. Calculates actual_minutes automatically."""
    # Non-admins can only create records for themselves
    if not current_user.is_admin and payload.user_id != current_user.id:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="他のユーザーの記録は作成できません")

    year_month = payload.work_date.strftime("%Y-%m")
    _assert_month_open(db, payload.user_id, year_month, current_user.is_admin)

    actual = calculate_actual_minutes(
        payload.start_time, payload.end_time, payload.break_minutes, payload.work_date
    )

    record = WorkRecord(
        **payload.model_dump(),
        actual_minutes=actual,
        created_by=current_user.id,
    )
    db.add(record)
    db.commit()
    db.refresh(record)
    return record


@router.get("/calendar/{user_id}/{year_month}", response_model=List[WorkRecordResponse])
def get_calendar_records(
    user_id: int,
    year_month: str,
    db: Session = Depends(get_db),
    current_user=Depends(get_current_user),
):
    """Return work records for a given user/month formatted for calendar display."""
    if not current_user.is_admin and current_user.id != user_id:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="アクセス権限がありません")

    year, month = year_month.split("-")
    from sqlalchemy import extract

    records = (
        db.query(WorkRecord)
        .filter(
            WorkRecord.user_id == user_id,
            extract("year", WorkRecord.work_date) == int(year),
            extract("month", WorkRecord.work_date) == int(month),
        )
        .order_by(WorkRecord.work_date)
        .all()
    )
    return records


@router.get("/{record_id}", response_model=WorkRecordResponse)
def get_work_record(
    record_id: int,
    db: Session = Depends(get_db),
    current_user=Depends(get_current_user),
):
    record = db.query(WorkRecord).filter(WorkRecord.id == record_id).first()
    if not record:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="記録が見つかりません")
    if not current_user.is_admin and record.user_id != current_user.id:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="アクセス権限がありません")
    return record


@router.put("/{record_id}", response_model=WorkRecordResponse)
def update_work_record(
    record_id: int,
    payload: WorkRecordUpdate,
    db: Session = Depends(get_db),
    current_user=Depends(get_current_user),
):
    record = db.query(WorkRecord).filter(WorkRecord.id == record_id).first()
    if not record:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="記録が見つかりません")
    if not current_user.is_admin and record.user_id != current_user.id:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="アクセス権限がありません")

    year_month = record.work_date.strftime("%Y-%m")
    _assert_month_open(db, record.user_id, year_month, current_user.is_admin)

    update_data = payload.model_dump(exclude_unset=True)
    for key, value in update_data.items():
        setattr(record, key, value)

    # Recalculate actual_minutes
    record.actual_minutes = calculate_actual_minutes(
        record.start_time, record.end_time, record.break_minutes, record.work_date
    )

    db.commit()
    db.refresh(record)
    return record


@router.delete("/{record_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_work_record(
    record_id: int,
    db: Session = Depends(get_db),
    current_user=Depends(get_current_user),
):
    record = db.query(WorkRecord).filter(WorkRecord.id == record_id).first()
    if not record:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="記録が見つかりません")
    if not current_user.is_admin and record.user_id != current_user.id:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="アクセス権限がありません")

    year_month = record.work_date.strftime("%Y-%m")
    _assert_month_open(db, record.user_id, year_month, current_user.is_admin)

    db.delete(record)
    db.commit()

File: backend/src/routes/workRecords.js
const express = require('express');
const router = express.Router();
const { db } = require('../database');
const { authMiddleware, adminMiddleware } = require('../auth');
const { calculateActualMinutes } = require('../services/calculation');

function isMonthClosed(yearMonth, userId) {
  const c = db.prepare("SELECT status FROM monthly_closings WHERE year_month=? AND user_id=?").get(yearMonth, userId);
  return c?.status === 'closed';
}

router.get('/', authMiddleware, (req, res) => {
  let { user_id, year_month } = req.query;
  if (!req.user.is_admin) user_id = req.user.id;
  let sql = 'SELECT wr.*, u.name as user_name FROM work_records wr JOIN users u ON wr.user_id=u.id WHERE 1=1';
  const params = [];
  if (user_id) { sql += ' AND wr.user_id=?'; params.push(user_id); }
  if (year_month) { sql += " AND wr.work_date LIKE ?"; params.push(`${year_month}-%`); }
  sql += ' ORDER BY wr.work_date, wr.start_time';
  res.json(db.prepare(sql).all(...params));
});

router.get('/calendar/:userId/:yearMonth', authMiddleware, (req, res) => {
  const { userId, yearMonth } = req.params;
  if (!req.user.is_admin && req.user.id !== parseInt(userId)) return res.status(403).json({ detail: '権限がありません' });
  const records = db.prepare("SELECT * FROM work_records WHERE user_id=? AND work_date LIKE ? ORDER BY work_date,start_time").all(userId, `${yearMonth}-%`);
  res.json(records);
});

router.post('/', authMiddleware, (req, res) => {
  const data = req.body;
  const yearMonth = data.work_date?.slice(0, 7);
  const userId = data.user_id || req.user.id;
  if (!req.user.is_admin && isMonthClosed(yearMonth, userId)) return res.status(400).json({ detail: 'この月は締め済みです' });
  const actual = calculateActualMinutes(data.start_time, data.end_time, data.break_minutes || 0);
  try {
    const info = db.prepare(`INSERT INTO work_records (user_id,work_date,start_time,end_time,break_minutes,actual_minutes,work_type,notes,created_by) VALUES (?,?,?,?,?,?,?,?,?)`)
      .run(userId, data.work_date, data.start_time, data.end_time, data.break_minutes || 0, actual, data.work_type || 'normal', data.notes || null, req.user.id);
    res.status(201).json(db.prepare('SELECT * FROM work_records WHERE id=?').get(info.lastInsertRowid));
  } catch (e) { res.status(400).json({ detail: e.message }); }
});

router.get('/:id', authMiddleware, (req, res) => {
  const r = db.prepare('SELECT * FROM work_records WHERE id=?').get(req.params.id);
  if (!r) return res.status(404).json({ detail: '記録が見つかりません' });
  if (!req.user.is_admin && req.user.id !== r.user_id) return res.status(403).json({ detail: '権限がありません' });
  res.json(r);
});

router.put('/:id', authMiddleware, (req, res) => {
  const r = db.prepare('SELECT * FROM work_records WHERE id=?').get(req.params.id);
  if (!r) return res.status(404).json({ detail: '記録が見つかりません' });
  if (!req.user.is_admin && req.user.id !== r.user_id) return res.status(403).json({ detail: '権限がありません' });
  const yearMonth = (req.body.work_date || r.work_date).slice(0, 7);
  if (!req.user.is_admin && isMonthClosed(yearMonth, r.user_id)) return res.status(400).json({ detail: 'この月は締め済みです' });
  const d = { ...r, ...req.body };
  const actual = calculateActualMinutes(d.start_time, d.end_time, d.break_minutes || 0);
  db.prepare(`UPDATE work_records SET work_date=?,start_time=?,end_time=?,break_minutes=?,actual_minutes=?,work_type=?,notes=?,updated_at=datetime('now','localtime') WHERE id=?`)
    .run(d.work_date, d.start_time, d.end_time, d.break_minutes, actual, d.work_type, d.notes, req.params.id);
  res.json(db.prepare('SELECT * FROM work_records WHERE id=?').get(req.params.id));
});

router.delete('/:id', authMiddleware, (req, res) => {
  const r = db.prepare('SELECT * FROM work_records WHERE id=?').get(req.params.id);
  if (!r) return res.status(404).json({ detail: '記録が見つかりません' });
  if (!req.user.is_admin && req.user.id !== r.user_id) return res.status(403).json({ detail: '権限がありません' });
  const yearMonth = r.work_date.slice(0, 7);
  if (!req.user.is_admin && isMonthClosed(yearMonth, r.user_id)) return res.status(400).json({ detail: 'この月は締め済みです' });
  db.prepare('DELETE FROM work_records WHERE id=?').run(req.params.id);
  res.json({ message: '削除しました' });
});

module.exports = router;
